src/legacy/server/saved_objects/service/lib/repository.ts

```diff
@@ -341,12 +342,12 @@ export class SavedObjectsRepository {
   /**
    * Deletes all objects from the provided namespace.
    *
-   * @param {string} namespace
+   * @param {Namespace} namespace
    * @returns {promise} - { took, timed_out, total, deleted, batches, version_conflicts, noops, retries, failures }
    */
-  async deleteByNamespace(namespace: string): Promise<any> {
-    if (!namespace || typeof namespace !== 'string') {
-      throw new TypeError(`namespace is required, and must be a string`);
+  async deleteByNamespace(namespace: Namespace): Promise<any> {
+    if (!namespace || !(namespace instanceof Namespace)) {
```

Comment:
Given the nature of TS, using `instanceof` might be problematic here... It's possible to use the following and have the type checking pass:

```
repository.deleteByNamespace({ id: 'some-namespace })
```

but then run into issues at run-time.

Reply:
You bring up an interesting point here...I feel like if we intend to allow for objects of shape `{ id?: string }` to be accepted, that we shouldn't be defining a `Namespace` class for this, but rather an interface.

Reply:
Yeah... I think ergonomically, using `new Namespace(string | undefined)` is better; where-as, using `{ id: string | undefined }` just looks awkward... I thought about suggesting that we use an interface for the methods which receive a namespace argument, and providing a class for ergonomic reasons, but we kinda get the best of both worlds as-is right now. It's super awkward though...

Reply:
hmmm...

The `createNamespace(id: string | undefined)` function on the spaces service could take care of ergonomics, and we could then update the methods which receive a namespace argument to use an interface.

BYO-Namespace should not be a very common task, even for heavy consumers of the saved objects service, so this might strike a good-enough balance

Reply:
That seems reasonable enough to me.